### PR TITLE
Minor improvements for package's components and tests

### LINKED ISSUE #183: Bring back 100% CC
After merging #178, the CC will drop below 100%.

### LINKED ISSUE #227: Watchdog causes memory leaks
☝️ Most probably the watchdog feature causes it. Whenever a new instance is created, the old reference is not destroyed (https://github.com/ckeditor/ckeditor5-react/blob/f271c4975ad9130bf9469b1fe797fdf886044597/src/ckeditor.jsx#L91-L95).

_Originally posted by @pomek in https://github.com/ckeditor/ckeditor5/issues/8725#issuecomment-802646255_

### PR DESCRIPTION
**Pull request by pomek**
### Suggested merge commit message ([convention](https://github.com/ckeditor/ckeditor5-design/wiki/Git-commit-message-convention))

Tests: Added the memory leak tests. Closes #227.

Tests: Added additional tests to cover all use-cases. Closes #183.

Tests: Added tests for dev-utils.

Tests: Grouped tests by test scenarios.

---

### Additional information

*For example – encountered issues, assumptions you had to make, other affected tickets, etc.*

### CODE REVIEW

File: tests/integrations/ckeditor-classiceditor.jsx [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	} );
+} );
+
+// The memory test based on: https://github.com/ckeditor/ckeditor5/blob/master/packages/ckeditor5-core/tests/_utils/memory.js.
+// It's the simplified, adjusted version that allows checking whether the <CKEditor> component destroys all references.
+const TEST_RETRIES = 2;
+const TEST_TIMEOUT = 5000;
+const GARBAGE_COLLECTOR_TIMEOUT = 500;
+
+describe( 'memory usage', () => {
```
Comment (pomek) [inline tests/integrations/ckeditor-classiceditor.jsx:44]:
```suggestion
describe( '<CKEditor> memory usage', () => {
```

File: src/ckeditorcontext.jsx [hunk 2]
```diff
@@ -13,34 +13,59 @@ export default class CKEditorContext extends React.Component {
...
+	/**
+	 * Re-render the entire component once again. The old editor will be destroyed and the new one will be created.
+	 */
+	componentDidUpdate() {
+		// If the `isLayoutReady` property has changed from `false` to `true`, the instance of `ContextWatchdog` does not exist.
+		if ( this.contextWatchdog ) {
+			this.contextWatchdog.destroy();
+		}
+
+		this._initializeContextWatchdog( this.props.config );
```
Comment (ma2ciek) [inline src/ckeditorcontext.jsx:56]:
```suggestion
		this._initializeContextWatchdog();
```

File: src/ckeditorcontext.jsx [hunk 3]
```diff
@@ -13,34 +13,59 @@ export default class CKEditorContext extends React.Component {
...
+
+	/**
+	 * Destroy the context before unmounting the component.
+	 */
+	componentWillUnmount() {
+		this._destroyContext();
+	}
+
+	/**
+	 * @return {JSX.Element}
```
Comment (ma2ciek) [inline src/ckeditorcontext.jsx:67]: TBH it's the React Component API. We don't need to document these methods :sweat\_smile:

**Comment by ma2ciek** [comment]
LGTM. Let's work on the code improvements in another PR :+1:

**Pull request merged**